### PR TITLE
Allow certain classes to be labeled as 'abstract'

### PR DESCRIPTION
**Pull request by phyy-nx**
An abstract dxtbx class is defined here as a class that should not be returned by the registry.

Mark FormatHDF5 as abstract

Fixes dials/dials#1220

This is an alternative to #254, marked as a draft here for discussion.  The implementation details could be different.  I tinkered with creating a decorator but didn't find a nice solution easily.  Regardless, with this convention, any class can be marked as abstract so it can't be found by the Registry.

Before these changes, using lyso009a_0087.JF07T32V01.mask.h5 from [zenodo](https://doi.org/10.5281/zenodo.3352357), a file that is understood by FormatHDF5 but is not an image file that should be understood by dxtbx, we got a gnarly exception, just like in dials/dials#1220:
```
$ dials.import lyso009a_0087.JF07T32V01.mask.h5
    <snip>
    assert single_file_indices is not None
```
With these changes, instead we get this:
```
$ dials.import lyso009a_0087.JF07T32V01.mask.h5
Sorry: Unable to handle the following arguments:
  lyso009a_0087.JF07T32V01.mask.h5
```

Inspecting `dxtbx.show_registry`, the following other classes could be marked as abstract: FormatMultiImage, FormatStill, and FormatXTC.

### CODE REVIEW

**Comment by Anthchirp** [comment]
Interesting.

However, I don't think the lookup should return `None` or anything other than the class (which is still the _correct_ leaf node format class), but the class itself should disallow unsubclassed instantiation with an exception.
IMO an exception saying "There is no format class underneath HDF5Format that wants to eat this file" will be a lot more useful than a "Sorry, can't handle argument".

I believe this would be possible to do with a decorator.

**Comment by phyy-nx** [comment]
Adding an assert to `__init__` that does a similar test (`assert type(self) is not FormatHDF5`) was my first thought as well but it turns out it doesn't fix dials/dials#1220.  The problem is that `assert single_file_indices is not None` is hit in `FormatMultiImage.get_imageset` before the format class is actually instantiated.  So I instead disabled it in the Registry.  That makes dials.import clean.

**Comment by phyy-nx** [comment]
(there still might be a better way but this is where I landed :)

**Comment by Anthchirp** [comment]
looking around there is already an `ignore()` class function that seems to serve a similar purpose, except that one declares formats that are supposed to be filtered out quitely.

**Comment by phyy-nx** [comment]
Yah, I used it as a model for is_abstract

**Comment by phyy-nx** [comment]
Ready for review

File: newsfragments/255.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Improve error message when attempting to import the wrong type of h5 file
```
Comment (graeme-winter) [inline newsfragments/255.feature:1]: ... attempting to import unsupported files

File: format/Registry.py [hunk 2]
```diff
@@ -119,10 +119,13 @@ def recurse(format_name, image_file):
...
     for format in sorted(_format_dag["Format"], key=format_sort):
         format_class = get_format_class_for(format)
         if scheme in format_class.schemes and format_class.understand(image_file):
-            return recurse(format, image_file)
+            result = recurse(format, image_file)
+            break
 
-    # There is no class accepting this file, so return None instead
-    return None
+    if result is not None and result.is_abstract():
```
Comment (graeme-winter) [inline format/Registry.py:129]: `if result and result.is_abstract()`

File: format/Format.py [hunk 3]
```diff
@@ -112,6 +112,10 @@ def understand(image_file):
     def ignore(cls):
         return False
 
+    @classmethod
+    def is_abstract(cls):
```
Comment (graeme-winter) [inline format/Format.py:116]: Possibly worth adding a docstring which explains what is meant by abstract in this case e.g. will be ignored by the registry if identified as the only supporting class

File: format/FormatHDF5.py [hunk 4]
```diff
@@ -8,6 +8,10 @@
 
 
 class FormatHDF5(FormatMultiImage, Format):
+    @classmethod
+    def is_abstract(cls):
```
Comment (graeme-winter) [inline format/FormatHDF5.py:12]: `is_abstract()` -> `abstract_format()` may be more in keeping? Proposal, not request

File: format/Format.py (hunk 3 above)
Comment (graeme-winter) [reply format/Format.py:116]: Also `is_abstract()` -> `abstract_format()` maybe?

**Review by graeme-winter — APPROVED** [review]
Assuming that this works (which I have not tested yet, but I assume is fine) this is a nicer solution to the alternative hack I proposed. I'd say a moderately verbose commit message when squashed & merged would be good as this will remain hidden for a long time until someone gets tripped up by it and needs to know what is going on...

**Comment by graeme-winter** [comment]
> Interesting.
> 
> However, I don't think the lookup should return `None` or anything other than the class (which is still the _correct_ leaf node format class), but the class itself should disallow unsubclassed instantiation with an exception.
> IMO an exception saying "There is no format class underneath HDF5Format that wants to eat this file" will be a lot more useful than a "Sorry, can't handle argument".
> 
> I believe this would be possible to do with a decorator.

If you could find a way to do this which is relatively general (i.e. not just limited to HDF5 files - I could see this being pertinent in the future) that would be neat _however_ this is still better than what we have now. Sorry I can't cope with these files: `kitten.h5` still better...

**Comment by graeme-winter** [comment]
@phyy-nx hang fire

My assertion that this is a good fix in review was not 100% correct viz:

```
Grey-Area x :) $ dials.import ../Therm_6_2_meta.h5
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 3.dev.272-geb1600a67
Sorry: Unable to handle the following arguments:
  ../Therm_6_2_meta.h5
```

This is much better!

```
Grey-Area x :( $ dials.import ../Therm_6_2_000001.h5
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
DIALS 3.dev.272-geb1600a67
The following parameters have been modified:

input {
  experiments = <image files>
}

Sorry: 
          Imageset contains no beam or detector model. This means you will be
          unable to process your data.

          Possible causes of this error are:
             - A problem reading the images with one of the dxtbx format classes
             - A lack of header information in the file itself.

          You can override this by specifying the metadata as geometry parameters
```

This is not helpful to our end user 🤔

Was what I was trying to avoid since this is the most common error case...

**Comment by graeme-winter** [comment]
BTW this is a manifestation of the HDF5-quagmire I was describing before. Additional:

```
Grey-Area x :) $ dxtbx.show_matching_formats ../Therm_6_2_000001.h5 |grep True
FormatHDF5: True
FormatRawData: True
```

This `RawData` one appears intentional & is not "abstract" as not inherited from:

```
grep FormatRawData ~/.../dxtbx/format/Format*
.../dxtbx/format/FormatHDF5RawData.py:class FormatRawData(FormatHDF5):
.../dxtbx/format/FormatHDF5RawData.py:        print(FormatRawData.understand(arg))
```

**Comment by phyy-nx** [comment]
Hm, both those error ￼￼messages look fine to me. The second message tells me something is wrong with the meta-data. In fact if you ran dials.import￼￼￼ and fully specify the geometry￼ on the command line you could probably process that data￼.

That said,￼￼ I think retiring FormatHDF5RawData is likely fine. Are you using it @dermen? Regardless, maybe that's not a blocker for this PR?

The code comments and suggestions are good. I'll put them in tomorrow (unless someone beats me to it :)

**Comment by phyy-nx** [comment]
Ooo I just remembered. I think FormatHDF5RawData can read a cheetah file if geometry is supplied. Dunno if there are folks doing that but at least it's a use case.

**Comment by graeme-winter** [comment]
> Ooo I just remembered. I think FormatHDF5RawData can read a cheetah file if geometry is supplied. Dunno if there are folks doing that but at least it's a use case.

Good candidate for moving to a special package?

**Comment by phyy-nx** [comment]
Code review added.

I also re-read FormatHDF5RawData and I see it only reads a single image, which is not a Cheetah file.  I think it can be safely deprecated.  I'll do that as a separate PR.

**Comment by phyy-nx** [comment]
I'ma merge this and #261 tomorrow unless there are more comments.  Thanks.

**Comment by Anthchirp** [comment]
could we discuss this in the core meeting maybe? I've only looked at this at the start when it was still marked 'draft here for discussion' and have since lost overview of how the PRs are connected

**Comment by phyy-nx** [comment]
> could we discuss this in the core meeting maybe? I've only looked at this at the start when it was still marked 'draft here for discussion' and have since lost overview of how the PRs are connected

Hi, these PRs aren't urgent.  Next core meeting it turns out is on a US holiday, but they could be discussed without me.  Regardless, it may suffice to say that #261 depends on this PR (#255), so this #255 goes first.

**Comment by graeme-winter** [comment]
> Hi, these PRs aren't urgent. Next core meeting it turns out is on a US holiday, but they could be discussed without me. Regardless, it may suffice to say that #261 depends on this PR (#255), so this #255 goes first.

If next core meeting on Thanksgiving, suggest move it forward by a day to Wednesday? @Anthchirp your thoughts? Idea is we have inclusive discussions 🙂

**Comment by ndevenish** [comment]
(I have some WIP changes that I accidentally pushed, still working on them)

**Comment by ndevenish** [comment]
It looks like a useful feature but I have a few suggestions, which since is nonzero work I've pushed myself as commit 5202904e7f7945b3af6d0f1137cd8f9d8491b826. I am happy to roll back if we don't like them.


- The mostly-disused `ignore()` feature should be removed, as I think this makes it completely redundant. I don't think it ever worked properly, so don't think that removing it is an API break for anyone
- A decorator seemed like the natural way to do it - abstract subclasses don't have to have the boilerplate properly in place
- I agree with Markus' point about not returning `None`, and failing instantiation - it makes the workings rather opaque. I've  replaced the places where `ignore()` was called to check `is_abstract`, and reverted the `Registry.py` changes - `None` is no longer returned, but the tests seem to pass now, `get_format` seems to work, and with the extra load handling `dials.import` seems to handle cleanly
- However, making `Format` fail to instantiate causes a whole slew of problems related to tests explicitly using `Format` (easy to fix), `check_format=False` with stills (harder, and [here](https://github.com/cctbx/dxtbx/blob/865909325b325f278ffafa2e864bd167edd0d889/format/Format.py#L359-L367), which I'm not convinced is doing the sensible thing with `check_format=False`) and the `test_FormatPYunspecified.py` tests failing (at which point I gave up and reverted the change). For that reason, I've allowed `Format` as an exception, which is a bit messy but gives us most of the protections - registry-wise this seems to be okay though, because `Format` will _never_ `understand()` a file
- Do we really like `is_abstract_format` more than `is_abstract`? It feels a little redundant, It's not kept here but happy to rename again. On that, I'm not totally sold on _abstract_ for the naming here, it feels a little, well, abstract. But I can't think of anything more descriptive that isn't mealy-mouthed, and it seems relatively consistent with the way we've used it elsewhere. So will put up with the discomfort.
- I've also added a test of the basic behaviour
- 🚲🏠 : `@abstract` or `@abstractformat` (like `@classmethod`). `is_abstract()` or `is_abstract_format()`?

`dials.import` still has the same result with the test file, which I think was agreed above as at least being "better":
```
$ dials.import lyso009a_0087.JF07T32V01.mask.h5
...
Sorry: Unable to handle the following arguments:
  lyso009a_0087.JF07T32V01.mask.h5
```

**Comment by phyy-nx** [comment]
Aweome.  Thanks @ndevenish for finishing this while I'm on beamtimes :)

**Pull request merged**